README.md
ePassportViewer
===============

This project has been developed by the Information Security Group (ISG) from the University of Louvain (ULC). It contains two parts, one is the python _pyPassport_ library that allows everyone to read his passport, and the GUI _ePassport Viewer_ that provides a friendly interface to the API. The tool allows everyone to check whether his passport resists to the attacks known in the literature.

For more information, you can read the paper from the ISG and al. published the in ACM journal: [A Survey of Security and Privacy Issues in ePassports](https://www.researchgate.net/publication/286047538_A_Survey_of_Security_and_Privacy_Issues_in_ePassports).


Installation guide
------------------

Here is the complete installation guide for __pyPassport__ and __ePassport Viewer__ on Linux (tested on Ubuntu 18.08 64bit) with an ACR122 using python 2.7.

First you need to download the latest version of ePassportViewer:

```
sudo apt install git
git clone https://github.com/beaujeant/ePassportViewer.git
```


### pyPassport

Once the repository cloned, you need to install the dependencies:

```
sudo apt install python
sudo apt install python-setuptools
sudo apt install python-crypto
sudo apt install python-pyasn1
sudo apt install python-pil
sudo apt install python-pyscard
```

You now can install pypassport:

```
cd ePassportViewer/pypassport
sudo python setup.py install
```


#### Optional

Some features of pypassport will only be available with the following additional dependencies:

```
sudo apt install openssl python-openssl
sudo apt install openjdk-6-jre-headless
```


### Install driver

First you need to unplug the reader, then install the following resource:

```
sudo apt install libusb-dev
sudo apt install pcscd
```

#### ACR

For the ACR122 from acs, you need to unplug the reader, then install the driver available here: https://www.acs.com.hk/en/driver/3/acr122u-usb-nfc-reader/

Here is example if you are on Ubuntu 18.08 64-bit:

```
cd /tmp/
wget https://www.acs.com.hk/download-driver-unified/10312/ACS-Unified-PKG-Lnx-116-P.zip
unzip ACS-Unified-PKG-Lnx-116-P.zip
cd ACS-Unified-PKG-Lnx-116-P/acsccid_linux_bin-1.1.6/ubuntu/bionic/
sudo dpkg -i libacsccid1_1.1.6-1~ubuntu18.04.1_amd64.deb
```

Once installed, you need to [unload the pn533](https://ludovicrousseau.blogspot.com/2013/11/linux-nfc-driver-conflicts-with-ccid.html):

```
sudo rm -r /lib/modules/*/kernel/drivers/nfc/pn533
```

Now you should be able to use your reader. You can connect your reader and test if the driver has been properly installed with the following command:

```
sudo service pcscd stop
sudo pcscd -f -d
# "Ctrl + C" to quit
```

When placing your passport on the reader, you should see something similar to:

```
00001241 eventhandler.c:404:EHStatusHandlerThread() powerState: POWER_STATE_POWERED
00000010 eventhandler.c:421:EHStatusHandlerThread() Card inserted into ACS ACR122U 00 00
00000008 Card ATR: 12 23 34 45 56 67 78 89 90 01 12 23 34 45 56 67 78 89 90 01
```


### ePassportViewer

ePassportViewer needs the following dependencies:

```
sudo apt install python-tk
sudo apt install python-imaging-tk
```

Once installed you can run `pcscd` service and start ePassportViewer:

```
sudo service pcscd start
cd ePassportViewer
python ePassportViewer.py
```

#### Optional

Some features of ePassportViewer will only be available with the following additional dependency:

```
sudo apt install libjpeg62
sudo apt install python-reportlab
```


### Troubleshooting

#### Failure to list reader

If you have the following error message:

```
'Failure to list readers: Service not Available.'.
Please check your passport is on the reader
```

Close ePassportViewer and run the following command:

```
sudo service pcscd restart
```


CHANGES.md
Changes
=======

From 2.0.6 to 2.1
-----------------

* Clean up the file structure
* Change documentation to Markdown format
* Update the _installation guide_ for Ubuntu 18.08 setup
* Fix PIL > Pillow lib


From 2.0.4c to 2.0.6
--------------------

ePassportViewer > mvc:

* New interface (menu + mrz entry)
* Clear function reset MRZ entry set black BAC in security frame
* Change color for security frame
* Move log menu
* Change configuration menu

ePassportViewer > dialog:

* Fix ABOUT > LICENCE issue
* Delete "OK" buttons
* Clean the "About" interface

ePassportViewer > attacks:

* Add a scroll bar in log

ePassportViewer > custom:

* Add HEX


From 2.0.3 to 2.0.4c
--------------------

ePassportViewer > frame > custom:

* Add the function 'Get ATR'

ePassportViewer > frame > attacks:

* Implementation of the verbose output

pyPassport > attacks > all:

* Fixing clumsy output

pyPassport > Iso7816 & ePassportViewer > frame > custom:

* Add method to get UID

pyPassport > fingerprint & ePassportViewer > mvc & ePassportViewer > dialog:

* Improving the fingerprint method.

ePassportViewer > dialog:

* Add ePassport generator features


From 2.0.2 to 2.0.3
-------------------

pyPassport > iso7816

* Modify rstConnection: set ciphering to False when reseting, which prevent unexpected encryption after reset.

ePassportViewer:

* Add geojasper.

pyPassport > jp2converter:

* Modification for geojasper location. Still clumsy and might raise error, but it makes things work with ePassportViewer.

ePassportViewer > dialog:

* Modify startReading: Handle reading problem because of EAC security.

* Create an dictionnary that append all DG in order to optimize the read of the ePassport for export functions.

ePassportViewer > frame > security:

* Set Active Authentication text in orange if DG15 can't be read.

pypassport > doc9303 > tagconverter:

* Create a new dictionnary to provide more info whenever a DG cannot be read because of EAC for instance.

ePassportViewer > mvc:

* Change menu File ("Save...", "Save as...").
* Improve exports methods (exprotToPDF and exportToXML) making more clear the path handling and optimizing the process.

ePassportViewer > util > inOut:

* Change function toPDF to provide more info

ePassportViewer > util > helper:

* Add new function to output raw dump of the passport


From 2.0.1 to 2.0.2
-------------------

pyPassport > setup.py

* The install didn't browse the sub directory. Now setup.py uses find_packages().

ePassportViewer > frame > attacks:

* Add the new methods addToHistory in order to keep the MRZ in history once used in attacks frame.

ePassportViewer > mvc:

* Improve add/load/save history methods in order to update MRZ added from the attacks frame.

pyPassport > epassport & pyPassport > attacks > all:

* Reset the connection before each communication in order to avoid 'Security status not satisfied' issues.


From 2.0 to 2.0.1
-----------------

pyPassport > attacks > bruteForce > BruteForce > offlineExploit:

* The offline brute force attack also checked if the the message decrypted match the nonce sent. Since it is not necessary and it is time consuming, the second verification has been removed.

pyPassport > attacks > bruteForce > BruteForce > initOffline:

* Since the nonce is not required anymore in the offline attack, the method return the message only.

ePassportViewer > ePassportViewer:

* The file is now executable.

ePassportViewer > frame > custom:

* Binary function 'AND', 'OR' and 'NOT' removed since they are not used.

ePassportViewer > frame > attacks:

* Add help dialogs for 'attacks' frame.
* The 'Reader #' is now in the menu. The 'French passport' check box is named 'Reach max'.
* Change the frame style of attacks in order to unify the style.


ePassportViewer/ePassportViewer.py
#!/usr/bin/python2.7

# Copyright 2012 Jean-Francois Houzard, Olivier Roger, Antonin Beaujeant
#
# This file is part of epassportviewer.
#
# epassportviewer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# epassportviewer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with epassportviewer.
# If not, see <http://www.gnu.org/licenses/>.

Error = False
import os
import tkMessageBox
try:
    from Tkinter import *
    from epassportviewer.const import *
    from epassportviewer.mvc import Controller
except Exception, msg:
    print msg
    tkMessageBox.showerror("Error", msg)
    Error = True

class dummyStream(object):
    ''' dummyStream behaves like a stream but does nothing. '''
    def __init__(self):
        self.f = open(OUTFILE, 'w')
    def write(self,data):
        self.f.write(data)
    def read(self,data): pass
    def flush(self):
        self.f.flush()
    def close(self):
        self.f.close()

def run():

    root = Tk()
    root.resizable(width = False, height = False)
    #TODO: Linux get error, windows not (missing app.ico)
    try:
        root.iconbitmap(os.path.join(os.path.dirname(os.path.realpath(__file__)), "app.ico"))
    except Exception, msg:
        pass
    # some files will be written temporarily in the current path:
    os.chdir(TMPDIR)

    app = Controller(parent=root)
    root.title(TITLE)
    root.mainloop()

if __name__ == "__main__":
    if not Error:
        run()


pypassport/pypassport/genpassport/datagroupcreation.py
from pypassport.doc9303 import converter, mrz
from pypassport.openssl import OpenSSL, OpenSSLException
from pypassport.logger import Logger
from pypassport.doc9303.datagroup import *
from pypassport.derobjectidentifier import *
from pypassport.asn1 import *

class DataGroupFileCreation(DataGroupFile):
    """
    The purpose of this class is to create a fake dataGroup.
    """
    def __init__(self, aid):
        """
        @param aid: The dataGroup tag
        @type aid: A string
        """
        DataGroupFile.__init__(self)
        self.tag = aid

    def addDataObject(self, tag, value):
        """
        Insert a new tag, value couple inside the dataGroup.

        @param tag: A tag in hexRep format
        @type tag: A string
        @param value: The value associated to the tag
        @type value: A string
        """
        length = binToHexRep(toAsn1Length(len(value)))
        tmp = hexRepToBin(tag) + hexRepToBin(length) + value
        self.body += tmp

    def getHeader(self):
        return hexRepToBin(self.tag) + toAsn1Length(len(self.body))

    def setHeader(self, value):
        super(DataGroupFileCreation, self).setHeader(value)

    header = property(getHeader, setHeader, None, None)

class Creation(Logger):
    def __init__(self):
        Logger.__init__("Creation")

    def create(self):
        raise Exception("Should be implemented")

    def _convertDate(self, date):
        if len(date) != 6:
            raise Exception("The date length is wrong")
        return date[4:6] + date[2:4] + date[0:2]

class ComCreation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("Common"))

    def create(self, dgs, LDSvn="30313037", Uvn="303430303030"):
        self._dgc.addDataObject("5F01", hexRepToBin(LDSvn))
        self._dgc.addDataObject("5F36", hexRepToBin(Uvn))

        dgsList = ""
        for dg in dgs:
            if dg.tag != '60' and dg.tag != '77':
                dgsList += str(dg.tag).upper()
        self._dgc.addDataObject("5C", hexRepToBin(dgsList))

        return Com(self._dgc).parse()

class DataGroup1Creation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("DG1"))

    def create(self, type, issuer, name, surname, nat, sex, passportID, birthDate, expiryDate):
        m = mrz.MRZ(None)
        forgedMRZ = m.buildMRZ(type, issuer, name, surname, nat, sex, passportID, self._convertDate(birthDate), self._convertDate(expiryDate))

        self._dgc.addDataObject("5F1F", forgedMRZ)

        return DataGroup1(self._dgc).parse()

class DataGroup2Creation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("DG2"))

    def create(self, imgPath):

        f = open(imgPath, 'rb')
        img = f.read()
        f.close()
        try:
            from PIL import Image
            width,height = Image.open(imgPath).size
        except:
            width = 0
            height = 0

        #Biometric Header Template
        bht = DataGroupFileCreation("A1")
        bht.addDataObject("87", hexRepToBin("0101"))
        bht.addDataObject("88", hexRepToBin("0008"))

        #Biometric Data Block
        bdb = DataGroupFileCreation("5F2E")
        bdb.body = ISO19794_5.createHeader('JPG', width, height, len(img)) + img

        #Biometric Information Group Template
        bigt = DataGroupFileCreation("7F61")
        bigt.addDataObject("02", hexRepToBin("01"))
        bigt.addDataObject("7F60", bht.file + bdb.file)

        self._dgc.body = bigt.file

        return DataGroup2(self._dgc).parse()

class DataGroup7Creation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("DG7"))

    def create(self, signPath):
        f = open(signPath, 'rb')
        self._dgc.addDataObject("02", hexRepToBin("01"))
        self._dgc.addDataObject("5F43", f.read())
        f.close()

        return DataGroup7(self._dgc).parse()

class DataGroup11Creation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("DG11"))

    def create(self, birthplace):
        self._dgc.addDataObject("5C", hexRepToBin("5F11"))
        self._dgc.addDataObject("5F11", birthplace)

        return DataGroup11(self._dgc).parse()

class DataGroup12Creation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("DG12"))

    def create(self, authority, issueDate):
        self._dgc.addDataObject("5C", hexRepToBin("5F195F26"))
        self._dgc.addDataObject("5F19", authority)
        self._dgc.addDataObject("5F26", self._convertDate(issueDate)) # DDMMYYYY => YYYYMMDD

        return DataGroup12(self._dgc).parse()

    def _convertDate(self, date):
        if len(date) != 8:
            raise Exception("The date length is wrong")
        return date[4:8] + date[2:4] + date[0:2]

class SODCreation(Creation):
    def __init__(self):
        self._dgc = DataGroupFileCreation(converter.toTAG("SecurityData"))
        self._hashAlgo = "sha1"
        self._openssl = OpenSSL()

    def create(self, ds, dsKey, dgs):
        sodContent = self._createSODContent(dgs)
        self._dgc.body = self._openssl.signData(sodContent, ds, dsKey)

        return SOD(self._dgc).parse()

    def _createSODContent(self, dgs):
        self._hashAlgo = "sha1"
        hashes = self._calculateHashes(dgs)

        lds = LDSSecurityObject()
        lds.setComponentByName('version', 0)

        ai = AlgorithmIdentifier()
        ai.setComponentByName('algorithm', ObjectIdentifier(OIDrevert[self._hashAlgo]))
        ai.setComponentByName('parameters', Null(''))
        lds.setComponentByName('hashAlgorithm', ai)

        dghv = DataGroupHashValues()

        keys = hashes.keys()
        keys.sort()
        cpt=0
        for hashNb in keys:
            dgh = DataGroupHash()
            dgh.setComponentByName('dataGroupNumber', Integer(hashNb))
            dgh.setComponentByName('dataGroupHashValue', OctetString(hashes[hashNb]))
            dghv.setComponentByPosition(cpt, dgh)
            cpt+=1

        lds.setComponentByName('dataGroupHashValues', dghv)

        return encoder.encode(lds)

    def _calculateHashes(self, dgs):
        hashes = {}
        hashAlgo = eval(self._hashAlgo)
        for dg in dgs:
            if dg.tag != '60' and dg.tag != '77':
                res = hashAlgo(dg.file)
                hashes[converter.toOrder(dg.tag)] = res.digest()

        return hashes


ePassportViewer/epassportviewer/frame/overview.py
# Copyright 2012 Jean-Francois Houzard, Olivier Roger, Antonin Beaujeant
#
# This file is part of epassportviewer.
#
# epassportviewer is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# epassportviewer is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public
# License along with epassportviewer.
# If not, see <http://www.gnu.org/licenses/>.

from Tkinter import *
import tkMessageBox
from PIL import Image, ImageTk
import tkFont

from string import replace

from epassportviewer.dialog import Tooltip
from epassportviewer.frame import security
from epassportviewer.const import *
from epassportviewer.util.configManager import configManager
from epassportviewer.util.image import convertImage, ImageFactory
from epassportviewer.util.helper import getItem

class Overview(Frame):

    IMAGE, SIGN = xrange(2)

    def __init__(self, master, callback):
        Frame.__init__(self, master, relief=GROOVE, bd=2)

        self.master = master
        self.callback = callback

        self.createLayout()

    def createLayout(self):
        container = Frame(self.master, relief=FLAT, bd=1)
        container.pack(side=TOP, fill=BOTH, expand=True)

        ##########
        ## MENU ##

        menuContainer = Frame(container)
        menuContainer.pack(side=LEFT, fill=Y)

        imRead = Image.open(ImageFactory().create(ImageFactory().READ))
        imageRead = ImageTk.PhotoImage(imRead)

        self.readButton = Button(menuContainer, image=imageRead, command=self.callback.process)
        self.readButton.image = imageRead
        self.readButton.pack(side=TOP, padx=3, pady=8)
        Tooltip(parent=self.readButton, tip="Read the passport")

        imAdditional = Image.open(ImageFactory().create(ImageFactory().ADDITIONAL))
        imageAdditional = ImageTk.PhotoImage(imAdditional)

        self.additionalButton = Button(menuContainer, image=imageAdditional, command=self.callback.AdditionalData, state=DISABLED)
        self.additionalButton.image = imageAdditional
        self.additionalButton.pack(side=TOP, padx=3)
        Tooltip(parent=self.additionalButton, tip="Get the entire content")

        imLog = Image.open(ImageFactory().create(ImageFactory().LOG))
        imageLog = ImageTk.PhotoImage(imLog)

        self.logButton = Button(menuContainer, image=imageLog, command=self.callback.openLog, state=DISABLED)
        self.logButton.image = imageLog
        self.logButton.pack(side=BOTTOM, padx=3, pady=8)
        Tooltip(parent=self.logButton, tip="Read the logs")


        ###################
        ## PROFILE PHOTO ##
        self.Image = Label(container)
        self.Image.pack(side=LEFT, pady=8, padx=10, anchor=N)

        im = Image.open( ImageFactory().create(ImageFactory().NOPICTURE) )
        Tkimage = ImageTk.PhotoImage(im)
        self.Image.config(image=Tkimage)


        ###################
        ## PERSONAL INFO ##

        dataContainer = Frame(container)
        dataContainer.pack(side=LEFT, fill=Y, padx=5, pady=5)

        titleFont = tkFont.Font(family="Helvetica", size=10)
        font = tkFont.Font(family="Helvetica", weight='bold', size=11)
        mrzFont = tkFont.Font(family='courier', size=14)

        self.fields = fields = {}
        fields['type'] = StringVar()
        fields['type'].set("N/A")
        fields['issueCountry'] = StringVar()
        fields['issueCountry'].set("N/A")
        fields['passportNumber'] = StringVar()
        fields['passportNumber'].set("N/A")
        fields['name'] = StringVar()
        fields['name'].set("N/A")
        fields['surname'] = StringVar()
        fields['surname'].set("N/A")
        fields['nationality'] = StringVar()
        fields['nationality'].set("N/A")
        fields['sex'] = StringVar()
        fields['sex'].set("N/A")
        fields['birthDate'] = StringVar()
        fields['birthDate'].set("N/A")
        fields['birthPlace'] = StringVar()
        fields['birthPlace'].set("N/A")
        fields['authority'] = StringVar()
        fields['authority'].set("N/A")
        fields['issueDate'] = StringVar()
        fields['issueDate'].set("N/A")
        fields['expiryDate'] = StringVar()
        fields['expiryDate'].set("N/A")
        fields['eyesColour'] = StringVar()
        fields['eyesColour'].set("N/A")
        fields['height'] = StringVar()
        fields['height'].set("N/A")
        fields['residence'] = StringVar()
        fields['residence'].set("N/A")

        fields['mrz'] = StringVar()


        # ROW 0
        row = 0
        Label(dataContainer, text="Type\t\t\t", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=2, sticky=W)
        Label(dataContainer, text="Issuing Country\t\t\t", font=titleFont, fg='blue').grid(row=row, column=2, columnspan=2, sticky=W)
        Label(dataContainer, text="Passport number", font=titleFont, fg='blue').grid(row=row, column=4, columnspan=2, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['type'], font=font).grid(row=row, column=0, columnspan=2, sticky=W)
        Label(dataContainer, textvariable=fields['issueCountry'], font=font).grid(row=row, column=2, columnspan=2, sticky=W)
        Label(dataContainer, textvariable=fields['passportNumber'], font=font).grid(row=row, column=4, columnspan=2, sticky=W)

        # ROW 2
        row += 1
        Label(dataContainer, text="Name", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=6, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['name'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)

        # ROW 4
        row += 1
        Label(dataContainer, text="Surname", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=6, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['surname'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)

        # ROW 6
        row += 1
        Label(dataContainer, text="Nationality", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, text="Height", font=titleFont, fg='blue').grid(row=row, column=6, columnspan=6, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['nationality'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, textvariable=fields['height'], font=font).grid(row=row, column=6, columnspan=6, sticky=W)

        # ROW 8
        row += 1
        Label(dataContainer, text="Date of birth", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, text="Place of birth", font=titleFont, fg='blue').grid(row=row, column=3, columnspan=6, sticky=W)
        Label(dataContainer, text="Colour of eyes", font=titleFont, fg='blue').grid(row=row, column=6, columnspan=6, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['birthDate'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, textvariable=fields['birthPlace'], font=font).grid(row=row, column=3, columnspan=6, sticky=W)
        Label(dataContainer, textvariable=fields['eyesColour'], font=font).grid(row=row, column=6, columnspan=6, sticky=W)

        # ROW 10
        row += 1
        Label(dataContainer, text="Sex", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, text="Authority", font=titleFont, fg='blue').grid(row=row, column=3, columnspan=6, sticky=W)
        Label(dataContainer, text="Residence", font=titleFont, fg='blue').grid(row=row, column=6, columnspan=6, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['sex'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)
        Label(dataContainer, textvariable=fields['authority'], font=font).grid(row=row, column=3, columnspan=6, sticky=W)
        Label(dataContainer, textvariable=fields['residence'], font=font).grid(row=row, column=6, columnspan=6, sticky=W)

        # ROW 12
        row += 1
        Label(dataContainer, text="Date of issue", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=3, sticky=W)
        Label(dataContainer, text="Signature", font=titleFont, fg='blue').grid(row=row, column=3, columnspan=3, sticky=W)
        row += 1

        self.Sign = Label(dataContainer)
        self.Sign.grid(row=row, rowspan=3, column=3, columnspan=3)

        im = Image.open( ImageFactory().create(ImageFactory().TRANSPARENT) )
        im.thumbnail((200,150))
        Tkimage = ImageTk.PhotoImage(im)

        self.Sign.config(image=Tkimage)
        self.Sign.image = Tkimage

        Label(dataContainer, textvariable=fields['issueDate'], font=font).grid(row=row, column=0, columnspan=6, sticky=W)
        row += 1

        # ROW 14
        Label(dataContainer, text="Date of Expiry", font=titleFont, fg='blue').grid(row=row, column=0, columnspan=3, sticky=W)
        row += 1
        Label(dataContainer, textvariable=fields['expiryDate'], font=font).grid(row=row, column=0, columnspan=3, sticky=W)


        #################
        ## MRZ LOG BOX ##

        mrzContainer = Frame(self.master, bd=2, relief=GROOVE)
        mrzContainer.pack(side=TOP, fill=BOTH, expand=True, anchor=CENTER)
        Label(mrzContainer, textvariable=fields['mrz'], bg='white', font=mrzFont).pack(side=LEFT, fill=BOTH, expand=True)


        ##############
        ## SECURITY ##

        self.security = security.securityFrame(self)
        self.security.pack(side=TOP, fill=BOTH, expand=True, anchor=CENTER)


        ##########
        ## PACK ##

        self.updatePicture( ImageFactory().create(ImageFactory().TRANSPARENT), self.SIGN)
        self.updatePicture( ImageFactory().create(ImageFactory().NOPICTURE), self.IMAGE)

        self.update()

    def updatePicture(self, path, type):
        im=Image.open(path)

        if type == self.SIGN:
            im.thumbnail((180,180))
        if type == self.IMAGE:
            im.thumbnail((275,275))

        Tkimage = ImageTk.PhotoImage(im)
        if type == self.IMAGE:
            self.Image.config(image=Tkimage)
            self.Image.image = Tkimage
        if type == self.SIGN:
            self.Sign.config(image=Tkimage)
            self.Sign.image = Tkimage

    def loadDG1(self, fields, data):
        if data.has_key('5F1F'):
            mrz = data['5F1F']
            mrz = mrz[:44] + "\n" + mrz[44:]
            if len(data['5F1F']) > 88:
                mrz = mrz[:89] + "\n" + mrz[89:]
            fields['mrz'].set(mrz)

        if data.has_key('5F5B'):
            name = (data['5F5B']).split("<<")
            fields['name'].set(getItem(name[0]))
            fields['surname'].set(getItem(name[1]))

        if data.has_key('5F03'):
            fields['type'].set(getItem(data['5F03']))
        if data.has_key('5F28'):
            fields['issueCountry'].set(getItem(data['5F28']))
        if data.has_key('5A'):
            fields['passportNumber'].set(getItem(data['5A']))
        if data.has_key('5F2C'):
            fields['nationality'].set(getItem(data['5F2C']))
        if data.has_key('5F57'):
            fields['birthDate'].set(getItem(data['5F57']))
        if data.has_key('5F35'):
            fields['sex'].set(getItem(data['5F35']))
        if data.has_key('59'):
            fields['expiryDate'].set(getItem("20" + data['59'])) # 20 make a "real" date

    def loadDG2(self, fields, data):
        tag = None
        if data['A1'].has_key('5F2E'): tag = '5F2E'
        elif data['A1'].has_key('7F2E'): tag = '7F2E'
        if tag != None:
            stream = convertImage(data['A1'][tag])
            try:
                self.updatePicture(stream, self.IMAGE)
            except Exception, msg:
                tkMessageBox.showwarning("Image Error", str(msg))

    def loadDG7(self, fields, data):
        if data.has_key('5F43'):
            try:
                stream = convertImage(data['5F43'][0])
                self.updatePicture(stream, self.SIGN)
            except Exception, msg:
                tkMessageBox.showwarning("Image Error", str(msg))

    def loadDG11(self, fields, data):
        if data.has_key("5F2B"):
            fields['birthDate'].set(getItem(data['5F2B']))
        if data.has_key("5F11"):
            fields['birthPlace'].set(getItem(data['5F11']))
        # This DG keep the accents, contrary to the DG1
        # but beware some passports contain an empty or almost empty 5F0E (e.g. BE: 5F0E0120 or 5F0E00)
        if data.has_key("5F0E") and \
           len(data['5F0E']) > len(fields['name'].get()) + len(fields['surname'].get()):
            if "<<" in data['5F0E']:
              tmpname = (data['5F0E']).split("<<")
              fields['name'].set(getItem(tmpname[0]))
              fields['surname'].set(getItem(tmpname[1]))
            else:
              # no specific separator between name(s) and surname(s)
              tmpname = (data['5F0E']).split(" ")
              namewords = len(fields['name'].get().split(" "))
              fields['name'].set(getItem(' '.join(tmpname[:namewords])))
              fields['surname'].set(getItem(' '.join(tmpname[namewords:])))
        # French, nom d'usage, http://fr.wikipedia.org/wiki/Nom_d'usage_en_France
        if data.has_key("A0") and getItem(data['A0'])[0] != '\x02':
            # TODO: some Belgian passports contain
            # 5F0E 00 A0 06020101 5F0F 00
            # is there a better way to discard their A0??
            old = fields['name'].get()
            fields['name'].set(old + ", " + getItem(data['A0']))


    def loadDG12(self, fields, data):
        if data.has_key("5F19"):
            fields['authority'].set(getItem(data['5F19']))
        if data.has_key("5F26"):
            fields['issueDate'].set(getItem(data['5F26']))

    def loadDG13(self, fields, data):
        if data.has_key("9F01"):
            fields['height'].set(getItem(data['9F01']) + " m")
        if data.has_key("9F02"):
            fields['eyesColour'].set(getItem(data['9F02']))
        if data.has_key("9F03"):
            fields['residence'].set(getItem(data['9F03']))

    def loadDG(self, dg, data):
        if dg in ["61", "67", "6B", "6C", "75", "6D"]:
            if dg == "61": self.loadDG1(self.fields, data)
            elif dg == "75": self.loadDG2(self.fields, data)
            elif dg == "67": self.loadDG7(self.fields, data)
            elif dg == "6B": self.loadDG11(self.fields, data)
            elif dg == "6C": self.loadDG12(self.fields, data)
            elif dg == "6D": self.loadDG13(self.fields, data)

        self.update()

    def clear(self):
        self.updatePicture( ImageFactory().create(ImageFactory().NOPICTURE), self.IMAGE)
        self.updatePicture( ImageFactory().create(ImageFactory().TRANSPARENT), self.SIGN)
        self.additionalButton.config(state=DISABLED)
        #self.logButton.config(state=DISABLED)
        for field in self.fields:
            self.fields[field].set("N/A")
        self.fields["mrz"].set("")
